express/manifest.py

```diff
@@ -1,112 +1,113 @@
-@dataclass
-class DataManifest:
+    @property
+    def fields(self) -> t.Dict[str, Field]:
+        return {
+            "id": Field(name="id", type="str"),
+            "source": Field(name="source", type="str"),
+        }
+
+
+class Manifest:
     """
-    The dataset Manifest
-    Args:
-        index (DataSource): the index parquet file which indexes all the data_sources
-        data_sources (List[DataSource]): Location and metadata of various data sources associated
-         with the index.
-        metadata (Metadata): The metadata associated with the manifest
+    Class representing an Express manifest
 
+    Args:
+        specification: The manifest specification as a Python dict
     """
 
-    index: DataSource
-    data_sources: Dict[str, DataSource] = field(default_factory=dict)
-    metadata: Metadata = field(
-        default_factory=Metadata
-    )  # TODO: make mandatory during construction
-
-    def __post_init__(self):
-        if (self.index.type != DataType.PARQUET) or (
-            not DataType.is_valid(self.index.type)
-        ):
-            raise TypeError("Index needs to be of type 'parquet'.")
-        for name, dataset in self.data_sources.items():
-            if not DataType.is_valid(dataset.type):
-                raise TypeError(
-                    f"Data type '{dataset.type}' for data source '{name}' is not valid."
-                )
+    def __init__(self, specification: dict) -> None:
+        self._validate_spec(specification)
+        self._specification = specification
+
+    @classmethod
+    def _validate_spec(cls, spec: dict) -> None:
+        """Validate a manifest specification against the manifest schema
+
+        Raises: InvalidManifest when the manifest is not valid.
+        """
+        spec_schema = json.loads(pkgutil.get_data("express", "schemas/manifest.json"))
+        validator = Draft4Validator(spec_schema)
+        try:
+            validator.validate(spec)
+        except jsonschema.exceptions.ValidationError as e:
+            raise InvalidManifest.create_from(e)
 
     @classmethod
-    def from_path(cls, manifest_path):
-        """Load data manifest from a given manifest path"""
-        with open(manifest_path, encoding="utf-8") as file_:
-            manifest_load = json.load(file_)
-            return DataManifest.from_dict(manifest_load)
+    def from_file(cls, path: str) -> "Manifest":
+        """Load the manifest from the file specified by the provided path"""
+        with open(path, encoding="utf-8") as file_:
+            specification = json.load(file_)
+            return cls(specification)
+
+    def to_file(self, path) -> None:
+        """Dump the manifest to the file specified by the provided path"""
+        with open(path, "w", encoding="utf-8") as file_:
+            json.dump(self._specification, file_)
+
+    @property
+    def metadata(self) -> dict:
+        return self._specification.get("metadata")
+
+    @property
+    def base_path(self) -> str:
```

Comment:
Rewrote this from scratch and only then saw that we were leveraging the bucket before instead of a general base path. If we use the bucket, we probably don't need to resolve the locations of the subsets.

@PhilippeMoussalli I think you have the best view on what approach would be most useful here.

Reply:
Before the path included both the protocol(e.g. gs://)+bucket+location (a.k.a blob). 

I would prefer to move away towards something where maybe we have a dataclass that separates all 3 of those? 

- Protocol: can be used to reconstruct the full path and identify the remote location of an object
- Bucket:  fuse takes a bucket and a mount path for mounting where the full mount path will be `mount_path= Path(mount_path/bucket)
- Location: used to track down the subset paths for reading/writing `Path(mount_path/location)` 

You could also just have the full path (e.g. gs://bucket/blob) and parse it to extract what you need from it. Not sure which approach is more suitable.

Reply:
I tried out this feature but it crashed my OS when running in background mode ([see doc](https://docs.google.com/document/d/1Yn8ayLBovv2PHlLQhsxp81Wzp8I23pOjW4xlMkqwmjs/edit#)). Also tried running it in the foreground with a thread and ran into similar issues. It's documented as an experimental feature but feel free to have a try it out. If you can make it work it will make things easier for us

Reply:
Does it provide a lot of benefit to split the protocol and the bucket? This is quite easy to do, right? Just thinking what would be easiest for the users.